ssl/quic/quic_route.c

```diff
@@ -0,0 +1,326 @@
+ */
+DEFINE_LIST_OF(remote_ids, QUIC_ROUTE);
+
+struct quic_route_table {
+    LHASH_OF(QUIC_ROUTE) *remote_id_tbl;
+    OSSL_LIST(remote_ids) remote_id_list;
+};
+
+/****************************************************************************
+ * Connection ID handling                                                   *
+ ****************************************************************************/
+
+/*
+ * Generate a new local connection ID
+ */
+int ossl_quic_conn_id_generate(QUIC_CONN_ID *cid)
+{
+    cid->id_len = QUIC_CONN_ID_LEN;
+    return RAND_bytes(cid->id, cid->id_len) > 0;
+}
+
+int ossl_quic_conn_id_set(QUIC_CONN_ID *cid, const unsigned char *id,
+                          unsigned int id_len)
+{
+    if (id_len > sizeof(cid->id))
+        return 0;
+#ifndef NDEBUG
+    if (id_len < sizeof(cid->id))
+        memset(cid->id + id_len, 0, sizeof(cid->id) - id_len);
+#endif
+    memcpy(cid->id, id, cid->id_len = id_len);
+    return 1;
+}
+
+/****************************************************************************
+ * Hash table lookup functions                                              *
+ ****************************************************************************/
+
+static unsigned long id_hash(const QUIC_CONN_ID *id)
+{
+    /*
+     * Naive implemenation of Bernstein's DJBX33A hash, starting with the length
+     * byte, followed by the ID itself.  For local ID's we could do better because
+     * we know the fixed length but it's unlikely to be worthwhile.
+     */
+    const unsigned int len = id->id_len;
+    unsigned long h = len;
+    unsigned int i;
+
+    for (i = 0; i < len; i++)
+        h = 33 * h + id->id[i];
+    return h;
+}
+
+static int id_cmp(const QUIC_CONN_ID *a, const QUIC_CONN_ID *b)
+{
+    if (a->id_len != b->id_len)
+        return a->id_len > b->id_len ? 1 : -1;
+    return memcmp(a->id, b->id, a->id_len);
+}
+
+static unsigned long local_id_hash(const QUIC_ROUTE *a)
+{
+    return id_hash(&a->local);
+}
+
+static int local_id_cmp(const QUIC_ROUTE *a, const QUIC_ROUTE *b)
+{
+    return id_cmp(&a->local, &b->local);
+}
+
+static unsigned long remote_id_hash(const QUIC_ROUTE *a)
+{
+    return id_hash(&a->remote);
+}
+
+static int remote_id_cmp(const QUIC_ROUTE *a, const QUIC_ROUTE *b)
+{
+    return id_cmp(&a->remote, &b->remote);
+}
+
+/****************************************************************************
+ * Local ID's (global)                                                      *
+ ****************************************************************************/
+
+static LHASH_OF(QUIC_ROUTE) *local_id_tbl;
+static CRYPTO_ONCE local_connection_id_init = CRYPTO_ONCE_STATIC_INIT;
+static CRYPTO_RWLOCK *local_connection_lock;
+
+static void local_connection_id_free(void)
+{
+    lh_QUIC_ROUTE_free(local_id_tbl);
+    local_id_tbl = NULL;
+    CRYPTO_THREAD_lock_free(local_connection_lock);
+    local_connection_lock = NULL;
+}
+
+DEFINE_RUN_ONCE_STATIC(local_connection_id_do_init)
```

Comment:
I really dislike this. I think we should be avoiding new run_once handlers. In fact I'd like to remove most of the ones we've currently got. "Global" stuff should be global to a libctx IMO. Although I'm unclear why this is global at all. Shouldn't we have a new set of local ids per SSL_CTX? The smaller the set of routes we have to manage means less lock contention

Reply:
We planned on vectoring all QUIC traffic through a single socket.  This kind of implies a global table of local ID to figure out which goes where.

Has the single socket assumption changed?

Reply:
I don't think we ever said QUIC traffic goes through a single socket *globally*. I would expect there to be a single socket per SSL_CTX. Two SSL_CTXs should be completely separate from each other.

Reply:
We were discussing promiscuous mode sockets at one point and I've raised this concern previously to silence.

It's easy enough to change to one local ID table per SSL_CTX.

Reply:
I think we should make that change